cni/pkg/ebpf/ambient_redirect_kern.c

```diff
@@ -0,0 +1,319 @@
+// clang  -O2  -Wall -target bpf -c ambient_redirect_kern.c -o ambient_redirect.o
+// POD_VETH=vethaab06c36
+// tc qdisc del dev ${POD_VETH} clsact
+// tc qdisc add dev ${POD_VETH} clsact
+// tc filter add dev ${POD_VETH} ingress bpf da obj ambient_redirect.o sec outbound
+// tc filter show dev ${POD_VETH} ingress
+// tc filter add dev ${POD_VETH} egress bpf da obj ambient_redirect.o sec inbound
+// tc filter show dev ${POD_VETH} egress
+// ZTU_VETH=veth526b5785
+// tc qdisc del dev ${ZTU_VETH} clsact
+// tc qdisc add dev ${ZTU_VETH} clsact
+// tc filter add dev ${ZTU_VETH}  ingress bpf da obj ambient_redirect.o sec  ztunnelhostingress
+// tc filter show dev ${ZTU_VETH}  ingress
+// NS=cni-4d297635-9c2d-78f0-3f3e-88b94f04cf5c
+// ip netns exec ${NS} tc qdisc add dev eth0 clsact
+// ip netns exec ${NS} tc filter add dev eth0 ingress bpf da obj ambient_redirect.o sec ztunnelingress
+
+#include <linux/bpf.h>
+#include <linux/pkt_cls.h>
+#include <netinet/in.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <linux/if_packet.h>
+#include <linux/if_ether.h>
+#include <linux/tcp.h>
+#include <stdint.h>
+#include <iproute2/bpf_elf.h>
+#include <bpf/bpf_helpers.h>
+#include <string.h>
+#include "common.h"
+
+
+#ifndef __section
+# define __section(NAME)                  \
+   __attribute__((section(NAME), used))
+#endif
+
+#ifndef __inline
+# define __inline                         \
+   inline __attribute__((always_inline))
+#endif
+
+
+#ifndef BPF_FUNC
+# define BPF_FUNC(NAME, ...)              \
+   (*NAME)(__VA_ARGS__) = (void *)BPF_FUNC_##NAME
+#endif
+
+static void *BPF_FUNC(map_lookup_elem, void *map, const void *key);
+static void BPF_FUNC(trace_printk, const char *fmt, int fmt_size, ...);
+
+#ifndef printk
+#define printk(fmt, ...)                                      \
+    ({                                                         \
+        char ____fmt[] = fmt;                                  \
+        trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
+    })
+#endif
+
+#define dbg(fmt, ...)                                      \
+    ({                                                         \
+        char ____fmt[] = fmt;                                  \
+        trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
```

Comment:
The performance of trace_printk is so low that we should consider not calling it if we are in non-debug mode.